auto_round/export/export_to_gguf/quant.py

```diff
@@ -0,0 +1,100 @@
+# Copyright (c) 2024 Intel Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import gguf
+import numpy as np
+
+QK_K = 256
+GGML_QUANT_SIZES = {
+    "bf16": (1, 2),
+    "q4_0": (32, 2 + 16),
+    "q4_1": (32, 2 + 2 + 16),
+    "q4_k": (256, 2 + 2 + QK_K // 2 + 12),
+}
+
+GGML_QUANT_BLOCK = {}
+
+def register_block(name):
+    def register(cls):
+        GGML_QUANT_BLOCK[name] = cls
+        return cls
+    return register
+
+
+def ggml_quant(data: np.array, ggml_type, scale = None, zp = None):
+    block_size, type_size = GGML_QUANT_SIZES[ggml_type]
+
+    data = data.astype(np.float32, copy=False)
+    shape = data.shape
+    n_blocks = data.size // block_size
+    blocks = data.reshape((n_blocks, block_size))
+    
+    new_data = GGML_QUANT_BLOCK[ggml_type](blocks, scale, zp)
+    new_data = new_data.reshape(*shape[:-1], shape[-1] // block_size * type_size)
+    return new_data
+
+@register_block("bf16")
+def bf16_quant_block(blocks: np.array, scale = None, zp = None):
+    n = blocks.view(np.uint32)
+    # force nan to quiet
+    n = np.where((n & 0x7fffffff) > 0x7f800000, (n & np.uint32(0xffff0000)) | np.uint32(64 << 16), n)
+    # round to nearest even
+    n = (np.uint64(n) + (0x7fff + ((n >> 16) & 1))) >> 16
+    return n.astype(np.uint16).view(np.uint8)
+
+@register_block("q4_0")
+def q4_0_quant_block(blocks: np.array, scale = None, zp = None):
+    if scale is not None:
```

Comment:
The symmetry of GGUF differs from what we are currently using., better have a test which one is better